Add dotnet-razor-precompile tool

pranavkm opened the pull request:
cc @rynowak \ @NTaylorMullen \ @davidfowl 

No strong naming as yet.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,153 @@
+                TargetFramework,
+                Configuration,
+                outputPath: outputPath,
+                buildBasePath: null,
```
The output path listed in xproj is actually used as the build-base-path. Super confusing I know. Even filed an issue but I guess it's by design? :/ https://github.com/aspnet/Tooling/issues/455

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,153 @@
+                dispatchArgs.Add("--generate-pdbs");
+            }
+
+            var toolName = typeof(Precompilation.Program).GetTypeInfo().Assembly.GetName().Name;
```
For inside-man binaries we're naming them `X.Design` for consistency

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,153 @@
+                .ExitCode;
+
+            var buildOutputPath = GetBuildOutputPath();
+            var updatedBinary = Directory.EnumerateFiles(buildOutputPath, $"*{ModifiedAssemblyExtension}").FirstOrDefault();
```
Why only taking the first?

pranavkm replied:
In theory there's only one we generate. Probably needs some to be hardened against stale binaries.

pranavkm replied:
Or better yet, look for the well known output name of the dll and use that.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,153 @@
+            }
+            else
+            {
+                File.Copy(
```
Why doesn't the inside man just override the DLL directly?

pranavkm replied:
We've assembly loaded it so consequently it's locked.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools.xproj:
```diff
@@ -1,20 +1,19 @@
   <PropertyGroup Label="Globals">
-    <ProjectGuid>6bb4c20b-24c0-45d6-9e4c-c2620959bdd5</ProjectGuid>
+    <ProjectGuid>f8bf7d95-0633-407f-bb0b-02563f13c068</ProjectGuid>
+    <RootNamespace>Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools</RootNamespace>
```
Nuke

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/project.json:
```diff
@@ -0,0 +1,51 @@
+    "Microsoft.DotNet.Cli.Utils": "1.0.0-*",
+    "Microsoft.Extensions.CommandLineUtils": "1.1.0-*",
+    "Microsoft.AspNetCore.Mvc.Razor.Precompilation": "1.0.0-*",
+    "Microsoft.Extensions.DotnetToolDispatcher.Sources": {
```
Only need this in netcoreapp1.0

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/AddResourcesRewriter.cs:
```diff
@@ -0,0 +1,70 @@
+                    IsPublic = true,
+                    Resource = new Resource
+                    {
+                        Data = output.AssemblyStream.ToArray().ToList(),
```
Why ToArray => ToList? Why not just ToList?

pranavkm replied:
`MemoryStream.ToArray()` and `Resource.Data` is of type `List<byte>`. Hence the gymnastics.

NTaylorMullen replied:
Makes me want to cry.

rynowak replied:
`List<byte>` is an extremely strange thing to put on a public API - is this part of the work-in-progress CCI stuff? Can you give them this feedback?

pranavkm replied:
Will do once I figure out what the repo is for this code.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/FIleProviderUtilities.cs:
```diff
@@ -0,0 +1,37 @@
+            return files;
+        }
+
+        private static void GetRazorFiles(IFileProvider fileProvider, List<RelativeFileInfo> razorFiles, string root)
```
Just inline.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/FIleProviderUtilities.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            foreach (var fileInfo in fileProvider.GetDirectoryContents(root))
+            {
+                var relativePath = Path.Combine(root, fileInfo.Name);
```
Kinda crazy to me that we don't expose the full path on the fileinfo.

NTaylorMullen replied:
Ah, nm. Lots of different file providers, wouldn't make sense always.

pranavkm replied:
I'm going to look into this issue. It's related to https://github.com/aspnet/Razor/issues/803 and I didn't want to change the semantics without figuring out the right thing to do.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompilationApplication.cs:
```diff
@@ -0,0 +1,58 @@
+            _callingType = callingType;
+
+            Name = "razor-tooling";
+            FullName = "Microsoft Razor Tooling Utility";
```
lmao

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompilationResult.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class PrecompilationResult
+    {
+        public string OutputPath { get; set; }
```
Hmm, you use this?

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileCommandBase.cs:
```diff
@@ -0,0 +1,184 @@
+                    ProjectPath = Path.GetDirectoryName(ProjectPath);
+                }
+
+                if (!Directory.Exists(ProjectPath))
```
Isn't this only possible if someone chooses to call the inside-man manually? I'd think the .Tools package would explode of the project.json wasn't valid.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileCommandBase.cs:
```diff
@@ -0,0 +1,184 @@
+
+        protected abstract int ExecuteCore();
+
+        private bool ParseArguments()
```
I feel like a lot of this code could exist only in the dispatcher bits. May be able to drop a dependency or two that way.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileCommandBase.cs:
```diff
@@ -0,0 +1,184 @@
+            }
+            else
+            {
+                Configuration = "Debug";
```
Constants.DefaultConfiguration

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileCommandBase.cs:
```diff
@@ -0,0 +1,184 @@
+            return true;
+        }
+
+        protected string GetBuildOutputPath()
```
You could pass the inside man this information so it didn't need to re-calculate it couldnt you?

pranavkm replied:
Sure, we could make it so that it always has all the values it needs.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            var services = new ServiceCollection();
+
+            var applicationName = Path.GetFileName(ProjectPath.TrimEnd('/'));
```
Directory separator

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,193 @@
+            var services = new ServiceCollection();
+
+            var applicationName = Path.GetFileName(ProjectPath.TrimEnd('/'));
+            var contentRoot = ContentRootOption.HasValue()
```
`ContentRootOption.Value() ?? Directory.GetCurrentDirectory()`

pranavkm replied:
Does that work for this type? It doesn't look like a value type.

NTaylorMullen replied:
It worked in previous tooling iterations. Did something change?

pranavkm replied:
Ahh, I thought `.Value()` threw when you didn't have a value. I can change this.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,193 @@
+                .AddSingleton<IHostingEnvironment>(hostingEnvironment)
+                .AddSingleton<DiagnosticSource>(diagnosticSource)
+                .AddLogging()
+                .AddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
```
Curious: Why we need this?

pranavkm replied:
Because we're terrible - https://github.com/aspnet/HttpAbstractions/issues/561

rynowak replied:
It's not that we're terrible, it's that there's no way to "register hosting services" that we can call, and our `AddXyz` aren't self-sufficient.

pranavkm replied:
The self sufficiency aspect of it is what makes it terrible.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,193 @@
+            var rewriters = new MetadataRewriter[]
+            {
+                new AddResourcesRewriter(host, precompilationResult.CompileOutputs),
+                new RemoveStrongNameRewriter(host)
```
lol

rynowak replied:
Other than the fact that we don't have the smarts to re-apply strong naming this is pretty clean.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            ParseArguments();
+
+            var services = ConfigureDefaultServices();
```
Abstract these two lines into an MvcRazorHostProvider. Too much going on in this class

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Microsoft.AspNetCore.Mvc.Razor.Precompilation.xproj:
```diff
@@ -4,17 +4,18 @@
   <PropertyGroup Label="Globals">
-    <ProjectGuid>a8aa326e-8ee8-4f11-b750-23028e0949d7</ProjectGuid>
+    <ProjectGuid>ebb55c82-6e61-4a0d-8ad8-86915696e86e</ProjectGuid>
+    <RootNamespace>Microsoft.AspNetCore.Mvc.Razor.Precompilation</RootNamespace>
```
nuke

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Program.cs:
```diff
@@ -0,0 +1,69 @@
+            return app.Execute(args);
+        }
+
+        private static void EnsureValidDispatchRecipient(ref string[] args)
```
I only did this in RazorTooling to limit the number of dependencies the inside man had. If you end up having to have cli.utils in your package graph for the inside man after the suggestions I've made above then you can use the DotnetToolDispatcher.EnsureValidDispatchRecipient

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using System.Reflection;
```
Boillller plate

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor/Compilation/IRoslynCompilationService.cs:
```diff
@@ -0,0 +1,28 @@
+    /// <summary>
+    /// A <see cref="ICompilationService"/> that uses Roslyn.
+    /// </summary>
+    public interface IRoslynCompilationService : ICompilationService
```
If we're going to start re-working how the CompilationService is talk with @cjqian on what she had to do to tie into the existing compilation service. Would rather ensure we have a coherent set of changes that enable VC TagHelpers and this.

pranavkm replied:
Yup. @davidfowl has some plans for a build manager like thing too.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs:
```diff
@@ -72,6 +72,11 @@ private static void AddRazorViewEngineFeatureProviders(IMvcCoreBuilder builder)
                 builder.PartManager.FeatureProviders.Add(new MetadataReferenceFeatureProvider());
             }
+
+            if (!builder.PartManager.FeatureProviders.OfType<PrecompiledViewsFeatureProvider>().Any())
```
Interesting, pre-compilation on by default?

pranavkm replied:
Yup, that's the plan. Since it technically is very low cost discovery, the hope is that it doesn't have a significant startup cost. The gesture to precompile is entirely optional.

tuespetre replied:
This is :boom: awesome

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/PrecompiledViewsFeatureProvider.cs:
```diff
@@ -0,0 +1,57 @@
+                    resourceName.EndsWith(".dll", StringComparison.Ordinal))
+                {
+                    var type = ReadResource(assembly, resourceName);
+                    var viewPath = resourceName.Substring(
```
resource names can have directory separators?

pranavkm replied:
Yup. Probably need to normalize the slashes as part of precompilation.

![image](https://cloud.githubusercontent.com/assets/174281/17234013/aeca0e7a-54e9-11e6-8fe0-0bdac3d311d0.png)

NTaylorMullen commented:
:watch:

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/project.json:
```diff
@@ -0,0 +1,51 @@
+  },
+  "dependencies": {
+    "Microsoft.AspNetCore.Hosting": "1.1.0-*",
+    "Microsoft.AspNetCore.Mvc": "1.1.0-*",
```
I'm guessing this doesn't depend on the metapackage. It would be nice to be more specific.

pranavkm replied:
Yup, working on removing some of these dependencies.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/CompileOutputs.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    public class CompileOutputs : IDisposable
+    {
+        public CompileOutputs(string relativePath, bool generatePdbs)
```
What's the scenario for not generating pdbs?

pranavkm replied:
Reducing the overall file size \ deployment size. We could make it an opt-out feature (as opposed to opt-in) if we care about it one way or the other.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/FIleProviderUtilities.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Internal
+{
+    public static class FileProviderUtilities
```
It's crazy that this isn't an extension method somewhere in file providers. Can we add it there instead?

pranavkm replied:
We do the weird thing where we try to preserve the relative path which makes it slightly hard to express this as a generic extension method. I could call this type something different if you'd like.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,193 @@
+            }
+
+            var compileOutputs = new CompileOutputs(fileInfo.RelativePath, GeneratePdbOption.HasValue());
+            var emitResult = CompilationService.EmitAssembly(
```
We'll discuss this at Monday's meeting

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation/Internal/RemoveStrongNameRewriter.cs:
```diff
@@ -0,0 +1,29 @@
+
+        public override IAssembly Rewrite(IAssembly assembly)
+        {
+            var mutableAssembly = base.Rewrite(assembly) as Assembly;
```
We might want to adopt the convention that I've seen in roslyn - use `CCI.Assembly` for these names. CCI tends to have a lot of overlap with type names in `System` or reflection

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Compilation/PrecompiledViewsFeature.cs:
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
```
This would probably not be `.Internal` btw.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompiledAssemblyUtility.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    public static class CompiledAssemblyUtility
```
gross

rynowak replied:
`AssemblyLoader`?

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompiledAssemblyUtility.cs:
```diff
@@ -0,0 +1,41 @@
+#else
+                System.Runtime.Loader.AssemblyLoadContext.Default.LoadFromStream(assemblyStream, pdbStream);
+#endif
+            return assembly.GetExportedTypes().FirstOrDefault(a => !a.IsNested);
```
Actually, I think this is really gross. I don't think this code is really reusable for anything other than what Razor does today (and it would be nice to improve what Razor does today).

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs:
```diff
@@ -196,7 +198,7 @@ private CSharpCompilation Rewrite(CSharpCompilation compilation)
             }
 
-            return compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(rewrittenTrees);
+            compilationContext.Compilation = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(rewrittenTrees);
```
I believe there is also a `ReplaceSyntaxTrees` which should save a step #ImmutabilityProblems

pranavkm replied:
There's a `ReplaceSyntaxTree` (singular) we could do, but ideally that would be part of the loop. The issue there is that we change the semantics of the rewrite ever so slightly (you'll see a rewritten tree if you are looking at other trees as part of the rewritter as opposed to the original tree).

i.e.

```
foreach (var tree in compilation.SyntaxTrees)
{
    var semanticModel = compilation.GetSemanticModel(tree, ignoreAccessibility: true);
    var rewriter = new ExpressionRewriter(semanticModel);

    var rewrittenTree = tree.WithRootAndOptions(rewriter.Visit(tree.GetRoot()), tree.Options);
    compilation = compilation.ReplaceSyntaxTree(tree, rewrittenTree);
}
```

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/PrecompiledViewsFeatureProvider.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public void PopulateFeature(IEnumerable<ApplicationPart> parts, PrecompiledViewsFeature feature)
+        {
+            foreach (var assemblyPart in parts.OfType<AssemblyPart>())
```
wrong

rynowak replied:
There should be an interface for this

pranavkm replied:
Do we want to consider doing the feature collection change prior to this? (https://github.com/aspnet/Mvc/issues/4760)? Seems really bad to add an interface every time we introduce new work here.

rynowak replied:
The problem isn't that we're adding a new interface, it's that they all have to be defined in `.Core`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/PrecompiledViewsFeatureProvider.cs:
```diff
@@ -0,0 +1,57 @@
+            foreach (var resourceName in resourceNames)
+            {
+                if (resourceName.StartsWith(PrecompiledResourcePrefix, StringComparison.Ordinal) &&
+                    resourceName.EndsWith(".dll", StringComparison.Ordinal))
```
you already have a constant for this

rynowak commented:
:watch: - this is overall looking like I expected. I don't grok the tools stuff so I'll let others comment on that :heart_eyes_cat:

pranavkm commented:
:up: :date

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/PrecompiledViewInfo.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public class PrecompiledViewInfo
+    {
+        public PrecompiledViewInfo(string path, Func<Stream> assemblyStreamFactory)
```
Why this overload? I only see you using the one below.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/PrecompiledViewInfo.cs:
```diff
@@ -0,0 +1,32 @@
+            Func<Stream> assemblyStreamFactory,
+            Func<Stream> pdbStreamFactory)
+        {
+            Path = path;
```
Should null check params

NTaylorMullen replied:
This comment probably applies everywhere in this PR

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs:
```diff
@@ -41,6 +51,32 @@ public AssemblyPart(Assembly assembly)
         /// <inheritdoc />
         public IEnumerable<TypeInfo> Types => Assembly.DefinedTypes;
 
+        public IEnumerable<PrecompiledViewInfo> PrecompiledViews
```
@rynowak thoughts on this?

rynowak replied:
We should figure out if we're moving view abstractions into MVC.Abstractions

At the **least** this change will unblock you for now.

rynowak replied:
Let's put a bookmark here to return to this at some point.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,239 @@
+                dispatchArgs,
+                TargetFramework,
+                Configuration,
+                outputPath: OutputPath,
```
https://github.com/aspnet/Mvc/pull/5087#discussion_r72674473

pranavkm replied:
Not entirely sure it applies here - we're not using the output path from the xproj. It's something the user optionally passes in.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/CommonOptions.cs:
```diff
@@ -0,0 +1,50 @@
+                "The path to the project (project folder or project.json) with precompilation.");
+
+            OutputPathOption = app.Option(
+                "-o|--output",
```
This might be changed by https://github.com/aspnet/Mvc/pull/5087/files#r73390069

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/PrecompiledViewInfo.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Microsoft.AspNetCore.Mvc.Core.ApplicationParts
+{
+    public class PrecompiledViewInfo
```
Just a note, not sure if there's anything that can be done but this is darn close to the `CompileOutput` API. Only difference is once has the assembly factory and one has the assembly.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/FIleProviderUtilities.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
+{
+    public static class FileProviderUtilities
```
Why does this need its own class? can't these methods be inlined in the one class that uses this?

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs:
```diff
@@ -0,0 +1,108 @@
+{
+    public class MvcServicesProvider
+    {
+        private const string ConfigureMvcMethod = "ConfigureMvc";
```
In Constants.cs?

NTaylorMullen replied:
Might make sense to expose this in Constants.cs as the default configure mvc method.

pranavkm replied:
Might be a question for @DamianEdwards, but would it be better if we had an interface that we discovered instead? There's the issue that we wouldn't be able to new it up, but are there other downsides to it? It does make the programming model for this a bit less magical.

rynowak replied:
Why is there a `Constants.cs`?

pranavkm replied:
I'm going to try the interface approach. EF has something similar so we have a bit of precedent around this. I'll run it by @DamianEdwards once he's back. If he feels strongly about keeping a static method we could revive the code.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs:
```diff
@@ -0,0 +1,108 @@
+            var serviceProvider = GetProvider(configureCompilationAction);
+
+            Host = serviceProvider.GetRequiredService<IMvcRazorHost>();
+            CompilationService = serviceProvider.GetRequiredService<ICompilationService>() as IRoslynCompilationService;
```
ew

pranavkm replied:
This will change once we have the BuildManager code @cjqian will write. I'll remove the interface, but we'll leave the cast for now).

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs:
```diff
@@ -0,0 +1,108 @@
+                    target: null);
+            }
+
+            // Todo: Add support for assembly scanning.
```
Like look for all reference assemblies?

pranavkm replied:
No, look for special attributes that tell us the method to invoke (in lieu of the command line argument).

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs:
```diff
@@ -0,0 +1,108 @@
+
+        private static Action<IMvcBuilder> GetConfigureCompilationAction(string configureCompilationType)
+        {
+            if (!string.IsNullOrEmpty(configureCompilationType))
```
```C#
if (string.IsNullOrEmpty(configureCompilationType))
{
    return null;
}
```

pranavkm replied:
We need to fill in the else eventually.

NTaylorMullen replied:
fair enough

NTaylorMullen commented:
:watch:

pranavkm commented:
More updates.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,207 @@
+                throw new ArgumentException("Project path not specified.");
+            }
+
+            if (!OutputFilePath.HasValue())
```
Why make this required? Can't we just use the default project/bin location if this isn't specified?

pranavkm replied:
Not really. This is the path at the runtime dir location. We'd have to recalculate it here. It's much easier to assume the tool has sorted this out and passed in the right value.

NTaylorMullen replied:
:ok:

NTaylorMullen commented:
Design looks good.

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -104,18 +80,13 @@ private int Execute()
                 Configuration,
-                outputPath: OutputPath,
-                buildBasePath: BuildBasePathOption.Value(),
+                outputPath: outputPaths.RuntimeOutputPath,
```
This works pretty well in the publish + post build scenario. Publish builds first to the default locations for a given configuration \ framework option. Prebuild is too early since no binaries exist at this stage and we'd be responsible for the build. This gives us a way out which works well for our primary scenario.

pranavkm commented on test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DefaultRoslynCompilationServiceTest.cs:
```diff
@@ -1,398 +1,398 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Will bring this back once we're ok with the design.

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/CSharpCompiler.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    public class CSharpCompiler
```
Placeholder types until @cjqian gets around to this.

NTaylorMullen replied:
Gonna skip reviewing this file then :smile:

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/CodeGenerator.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
+{
+    public class CodeGenerator
```
Name  \ design suggestions?

NTaylorMullen replied:
`ViewCollectionInjector` ?

pranavkm commented:
:up: :date:

We decided to move away from modifying generated assemblies to producing assemblies with a name derived from the primary assembly. Additionally the primary scenario here is to simplify the publish workflow and consequently I removed support for building as part of precompilation. It's assumed that the project is built by the time this tool is called.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs:
```diff
@@ -41,6 +51,25 @@ public AssemblyPart(Assembly assembly)
+                var precompiledAssemblyName = new AssemblyName(Assembly.FullName);
+                precompiledAssemblyName.Name = precompiledAssemblyName.Name + PrecompiledViewsAssemblySuffix;
+
+                var viewFactoryTypeName = $"{ViewFactoryNamespace}.{ViewFactoryTypeName},{precompiledAssemblyName}";
```
Odd to call it a factory when its really just a collection

pranavkm replied:
Sure. I was initially was playing around with trying to codegen the type factory we create in https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCacheResult.cs#L47 which is why I came up with this name. We could still do that here.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -178,30 +192,37 @@ private static void GetRazorFiles(IFileProvider fileProvider, List<RelativeFileI
+                containingNamespace = containingNamespace.ContainingNamespace;
             }
+
+            nameBuilder.Append(typeModel.MetadataName);
```
Curious: This understands generic type names well?

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,228 @@
+                containingNamespace = containingNamespace.ContainingNamespace;
+            }
+
+            nameBuilder.Append(typeModel.MetadataName);
```
Curious: This understands generic type names well?

pranavkm replied:
It doesn't, but we're only ever dealing with Razor view types and the view engine pipeline does not support instantiating generic types.

NTaylorMullen replied:
Was just curious :smile:

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -178,30 +192,37 @@ private static void GetRazorFiles(IFileProvider fileProvider, List<RelativeFileI
+        }
 
-            public PdbWriter PdbWriter { get; }
+        private string GetFullName(INamedTypeSymbol typeModel)
```
Use `ToDisplayString`

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/CodeGenerator.cs:
```diff
@@ -0,0 +1,68 @@
+            Compilation = Compilation.AddSyntaxTrees(syntaxTree);
+        }
+
+        public void AddAssemblyMetadata(AssemblyName applicationAssemblyName, string keyFilePath)
```
Looks like we're only ever passing in null here from call sites. Is that intentional?

pranavkm replied:
We need to pass the key file through.

NTaylorMullen replied:
Is that just not hooked up in the PR yet?

pranavkm replied:
Yup. It'll come from the dispatch command (since we've already parsed the project file).

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,228 @@
+            Options.Configure(app);
+            OutputPathOption = app.Option(
+                "--output-path",
+                "Path to the emit the precompiled assembly in.",
```
`Path to emit the precompiled assembly to.`

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/ViewCompilationInfo.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
+{
+    public class ViewCompilationInfo
```
This type is being used for two different purposes. 1 parsing a view and then determining errors at parse time. 2 expecting the GeneratorResults to be compiled and TypeName extracted so that this can be passed to the `ViewCollectionGenerator` (currently `CodeGenerator`).

pranavkm replied:
Isn't this just a container for stuff used for complication?

NTaylorMullen replied:
Maybe I'm just too used to them being separate beings hah. Fair enough.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Program.cs:
```diff
@@ -0,0 +1,67 @@
+        public static int Main(string[] args)
+        {
+#if DEBUG
+            DebugHelper.HandleDebugSwitch(ref args);
```
`[Condition]` is your friend :smile:

pranavkm replied:
I don't want to define the type at all in release builds. AFAIK, you can't stick that attribute on a type.

NTaylorMullen replied:
Well that's sad.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,164 @@
+            var applicationName = Path.GetFileNameWithoutExtension(outputPaths.CompilationFiles.Assembly);
+            var dispatchArgs = new List<string>
+            {
+                "--debug",
```
Not intentional right?

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,164 @@
+                "--application-name",
+                applicationName,
+                "--output-path",
+                GetOutputPath(),
```
Pull into its own variable so it's easy to debug if needed.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,164 @@
+                TargetFramework,
+                Configuration,
+                outputPath: outputPaths.RuntimeOutputPath,
+                buildBasePath: null,
```
Curious why this was changed.

pranavkm replied:
We calculate the output path which is where publish would have dropped the output to. We don't need the build base path in this case.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -174,69 +134,31 @@ private OutputPaths GetOutputPaths(IEnumerable<string> runtimeIdentifiers)
-                return projectContexts[0].TargetFramework;
-            }
-            else
+        private string GetOutputPath()
```
Would prefer to stay consistent with the Design pattern where you have a single ParseArguments that validates the inputs to the application.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs:
```diff
@@ -0,0 +1,164 @@
+
+        private string GetOutputPath()
+        {
+            if (!OutputPathOption.HasValue())
```
Would prefer to stay consistent with the Design pattern where you have a single `ParseArguments` method to validate input into the program.

pranavkm replied:
Sure.

NTaylorMullen commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorReferenceManager.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    public class RazorReferenceManager
```
ditto

NTaylorMullen commented:
:watch:

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs:
```diff
@@ -41,6 +51,25 @@ public AssemblyPart(Assembly assembly)
         /// <inheritdoc />
         public IEnumerable<TypeInfo> Types => Assembly.DefinedTypes;
 
+        public PrecompiledViews PrecompiledViews
```
I have some ideas about how to improve the factoring of this in the future, but let's leave that for another PR.

pranavkm commented:
Added support for strong name signing + addressed PR comments. Moving this out of design. I'll work on cleaning up the test that was commented out. Tests for this would have to go to a separate repo. I'll work with @Eilon on this.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs:
```diff
@@ -5,15 +5,24 @@
     {
+        public static readonly string PrecompiledViewsAssemblySuffix = ".PrecompiledViews";
+        public static readonly string ViewCollectionNamespace = "AspNetCore";
+        public static readonly string ViewCollectionTypeName = "__PrecompiledViewCollection";
```
Is there a reason not to just make these two properties this a single string?

pranavkm replied:
I use these in codegen, so they have to separated.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs:
```diff
@@ -42,6 +51,27 @@ public AssemblyPart(Assembly assembly)
+                    return null;
+                }
+
+                var precompiledViews = Activator.CreateInstance(viewFactoryType) as PrecompiledViews;
```
Seems like this should be a c-style cast. You want to get an exception if this fails.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs:
```diff
@@ -42,6 +51,27 @@ public AssemblyPart(Assembly assembly)
         public IEnumerable<TypeInfo> Types => Assembly.DefinedTypes;
 
         /// <inheritdoc />
+        public IReadOnlyCollection<PrecompiledViewInfo> PrecompiledViews
```
Should we do `IEnumerable` or `IReadOnlyList` here? We don't really use `IReadOnlyCollection` anywhere else, and there isn't a reason to prefer it over `IEnumerable`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/IPrecompiledViewsProvider.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <summary>
+    /// Exposes a sequence of precompiled views associated with an <see cref="ApplicationPart"/> .
+    /// </summary>
+    public interface IPrecompiledViewsProvider
```
Can we drop the term `precompiled`? That really only makes sense to us. I think a lot of these APIs would be cleaner and more general without `precompiled` in the names.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/IPrecompiledViewsProvider.cs:
```diff
@@ -0,0 +1,19 @@
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+
+namespace Microsoft.AspNetCore.Mvc.Core.ApplicationParts
```
Drop the `.Core`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/PrecompiledViewInfo.cs:
```diff
@@ -0,0 +1,34 @@
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc.Core.ApplicationParts
```
Drop the core

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/PrecompiledViews.cs:
```diff
@@ -0,0 +1,24 @@
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.Core.ApplicationParts
```
Get rid of `.Core`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/PrecompiledViews.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// A container for <see cref="PrecompiledViewInfo"/> instances.
+    /// </summary>
+    public abstract class PrecompiledViews
```
Naming seems bad. should this be something `ViewFeature`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/DesignTime/IMvcBuilderConfiguration.cs:
```diff
@@ -0,0 +1,20 @@
+    /// Configures the <see cref="IMvcBuilder"/>. Implement this interface to enable design-time configuration
+    /// (for instance during pre-compilation of views) of <see cref="IMvcBuilder"/>.
+    /// </summary>
+    public interface IMvcBuilderConfiguration
```
Do we need the equivalent for `IMvcCoreBuilder`?

rynowak replied:
Or does that get too complex.

pranavkm replied:
It might not be entirely too difficult. We could prefer types implementing `IMvcBuilderConfiguration` over `IMvcCoreBuilderConfiguration`.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/DesignTime/IMvcBuilderConfiguration.cs:
```diff
@@ -0,0 +1,20 @@
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.DesignTime
```
I'm not sure another namespace is justified.

pranavkm replied:
Should we name the type `IMvcDesignTimeBuilderConfiguration` in that case? Would be nice to call out that the interface is specifically meant for design time.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/DesignTime/IMvcBuilderConfiguration.cs:
```diff
@@ -0,0 +1,20 @@
+        /// Configures the <see cref="IMvcBuilder"/>.
+        /// </summary>
+        /// <param name="builder"></param>
+        void ConfigureMvc(IMvcBuilder builder);
```
since this is an interface method, is the idea that you'd put this on your startup class? Or do you do another class?

Thinking about code sharing between designtime and runtime

pranavkm replied:
Yeah, in my sample, I added it to Startup. It made it more natural to take the code you had in `ConfigureServices` and move it into this one.

pranavkm replied:
See https://github.com/pranavkm/MvcPrecompilation/pull/1/files#diff-9a30a8c3e2e7e199409cca66c004b6bfR14

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/CommonOptions.cs:
```diff
@@ -0,0 +1,39 @@
+
+using Microsoft.Extensions.CommandLineUtils;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
```
I'm largely ignoring the stuff to do with the command line application - just FYI

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/DebugHelper.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
+{
+    public static class DebugHelper
```
![image](https://cloud.githubusercontent.com/assets/1430011/17753773/2d760c78-6486-11e6-996a-9c368348d2a4.png)

pranavkm replied:
I kept expecting it play...

NTaylorMullen replied:
I can second that.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs:
```diff
@@ -0,0 +1,115 @@
+            var mvcBuilder = new MvcBuilder(mvcCoreBuilder.Services, mvcCoreBuilder.PartManager);
+            mvcBuilderConfiguration?.ConfigureMvc(mvcBuilder);
+
+            return mvcBuilder.Services.BuildServiceProvider();
```
/cc @davidfowl - this is what we discussed a while ago about our `AddXyZ(...)` methods not being self-sufficient. This is necessary every where we need to use MVC with DI, but not with hosting.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompilationApplication.cs:
```diff
@@ -0,0 +1,61 @@
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine(ex.Message);
```
Why hide the stack trace? If a customer reports a bug that's going to be the first thing we ask them for.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,238 @@
+
+        private CommandOption ApplicationNameOption { get; set; }
+
+        private MvcServicesProvider ServicesProvider { get; set; }
```
Why is this `ServicesProvider` - don't add random plurals to things

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,238 @@
+            }
+        }
+
+        private ViewCompilationInfo[] ParseViews()
```
This isn't really parsing, this is parsing and code generation

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/PrecompileRunCommand.cs:
```diff
@@ -0,0 +1,238 @@
+            var classDeclarations = syntaxTree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>();
+            foreach (var declaration in classDeclarations)
+            {
+                var typeModel = semanticModel.GetDeclaredSymbol(declaration);
```
`typeSymbol`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Razor/Internal/CSharpCompiler.cs:
```diff
@@ -0,0 +1,81 @@
+                references: _referenceManager.CompilationReferences);
+        }
+
+        public CSharpCompilation ProcessCompilation(CSharpCompilation compilation)
```
This shouldn't be here - this is specific to Razor compilation

rynowak commented:
:shipit: - Let's get this in and address issues in follow ups

pranavkm commented:
Moving the Mvc specific changes to a separate commit: https://github.com/aspnet/Mvc/pull/5160. I'll close this once I've addressed PR comments for the tool specific changes.
